lib/codecs/UrlCodec.js

```diff
@@ -100,12 +103,45 @@ export class Base58DidUrlCodec {
     return decodedValue;
   }
 }
+const uuidRegex = /(urn:uuid):(.*)/g;
+export class UrnUuidCodec {
+  encode(value) {
+    // break the base58 DID URL into segments
+    const matches = Array.from(value.matchAll(uuidRegex));
+    const match = matches[0];
+
+    // encode the DID URL segments
+    const encodedUrl = (urlHeaders.has(match[1])) ? urlHeaders.get(match[1]) :
+      match[1];
+    // FIXME: Assumes all UUIDs are lowercase, which will preserve the value
+    //        but might destroy signatures. Do not compress unless the UUIDs
+    //        are all lowercase?
```

Comment:
@dlongley can you pls explain more what needs to be done here? I'm not sure I understand how to resolve this.

Reply:
My comment is in response the FIXME question about what to do here if UUIDs have uppercase characters in them. If they do, that information would be lost when they are compacted. That could cause digital signatures to fail; therefore, we need the encoding scheme to either keep track of that information (no, we're not doing that) or we should only apply the encoding when there are no uppercase characters.

So, I'm not sure how these codecs work, but *either*:

1. We need a way for them to say "don't encode, just keep the value the same" (maybe that is accomplished simply by returning `value` as-is, maybe not), OR
2. We'll have to detect that the codec doesn't apply somewhere else and not call this codec function, OR
3. We're supposed to throw an exception when the UUID has uppercase characters because it's the user's responsibility not to hook up the codec in that case.

I don't know which of these is the right answer. I'd say that you should have a look at what the other codecs are doing, understand how this library is called/used, and try and figure out what the best path forward here is.

The end result needs to be that if a UUID has uppercase characters, it doesn't get encoded but is instead treated as a string.

Reply:
I *suppose* we could use yet another byte if the UUIDs have *all* uppercase characters -- and follow a similar encoding scheme to what we did with sub-millisecond precision for dates. We'd still have to throw if there were a mix of uppercase/lowercase though. That's probably the way to go unless you find a better path.

Reply:
@JSAssassin -- Ok, here's what we want to do. We should check the value to see if it's all lowercase. If it is NOT, then just return the value as is. We will need a test to make sure this will round trip properly (we can encode `urn:uuid:UPPERCASESTUFF` and decode it again without issue).

Then, if it is all lowercase, then we do what the rest of the code here is doing and compress the value down. We'll also need to ensure the test round trips that version.

Finally, we should have a test that has both uppercase and lowercase in the same document and ensure it round trips. Then we're done. Does that sound good to you?